src/crimson/osd/main.cc

```diff
@@ -38,54 +38,55 @@ seastar::logger& logger() {
+auto partition_args(int argc, const char *argv[])
 {
-  // collect all options consumed by seastar::app_template
-  auto parsed = bpo::command_line_parser(std::distance(argv_begin, argv_end),
-                                         argv_begin)
-    .options(app.get_options_description())
-    .style(bpo::command_line_style::default_style &
-           ~bpo::command_line_style::allow_guessing)
-    .allow_unregistered().run();
-  auto unknown_args = bpo::collect_unrecognized(parsed.options,
-                                                bpo::include_positional);
-  std::vector<const char*> ceph_args, app_args;
-  // ceph_argparse_early_args() and
-  // seastar::smp::get_options_description() use "-c" for different
-  // options. and ceph wins
-  auto consume_conf_arg = [&](char** argv) {
-    if (std::strcmp(*argv, "-c") == 0) {
-      std::cout << "warn: apply '-c FILE' as ceph option" << std::endl;
-      ceph_args.push_back(*argv++);
-      if (argv != argv_end) {
-        ceph_args.push_back(*argv++);
-      }
-    }
-    return argv;
-  };
-  auto unknown = unknown_args.begin();
-  auto consume_unknown_arg = [&](char** argv) {
-    for (; unknown != unknown_args.end() &&
-           argv != argv_end &&
-           *unknown == *argv; ++argv, ++unknown) {
-      ceph_args.push_back(*argv);
+  std::vector<const char *> app_args = argv_to_vec(argc, argv);
+  assert(app_args.size() > 0);
+
+  std::vector<const char *> ceph_args;
+  const char* program_name = app_args[0];
+  ConfigValues values;
+
+  class DummyTracker : public ConfigTracker {
+    bool is_tracking(const std::string& name) const override {
+      return false;
     }
-    return argv;
   };
-  for (auto argv = argv_begin; argv != argv_end;) {
-    if (auto next_arg = consume_conf_arg(argv); next_arg != argv) {
-      argv = next_arg;
-    } else if (auto next_arg = consume_unknown_arg(argv); next_arg != argv) {
-      argv = next_arg;
-    } else {
-      app_args.push_back(*argv++);
-    }
+  DummyTracker dt;
+  md_config_t config(values, dt, true);
+
+  // first copy what's in app_args into ceph_args
+  for (auto it = begin(app_args); it != end(app_args);) {
+    ceph_args.emplace_back(*it++);
   }
+
+  // pull off the stuff from app_args that we'll give to local_conf
+  config.parse_argv(values, dt, app_args, CONF_CMDLINE);
+
+  // Now that we have a filtered app_args, strip those values from ceph_args
+  [&]() {
+    for (auto it = begin(ceph_args); it != end(ceph_args);) {
+      for (auto it2 = begin(app_args); it2 != end(app_args);) {
+        if (it != end(ceph_args) && *it == *it2) {
+          it = ceph_args.erase(it);
+          if (it == end(ceph_args)) {
+            return;
+          }
+          it2 = begin(app_args);
+        } else {
+          ++it2;
+        }
+      }
+      ++it;
+    }
+  }();
+
+  // finally, copy the program name back into app_args for seastar
+  app_args.emplace(begin(app_args), program_name);
```

Comment:
sine the whole point of creating `DummyTracker` and `ConfigValues` is to create `app_args`, i'd suggest enclose them in a lambda for better readability, like
```c++
  auto app_args = [=] {
    class DummyTracker : public ConfigTracker {
        bool is_tracking(const std::string& name) const override {
          return false;
        }
    };
    DummyTracker dt;
    ConfigValues config_values;
    md_config_t config(config_values, dt, true);
    std::vector<const char *> app_args{argv + 1, argv + argc};
    // pull off the stuff from app_args that we'll give to local_conf
    config.parse_argv(config_values, dt, app_args, CONF_CMDLINE);
    return app_args;
  }();
```

Reply:
Yeah, I like this.  Will adopt with a couple of small changes.